### PR TITLE
Decrease the minSdk to 26

### PR DESCRIPTION
**Pull request by wh201906**
This PR replicates the changes in #389 but is updated with the latest commit (adb0175abb37aae37332a586b76848fa9f58ad56) from `main`.

### CODE REVIEW

**Comment by wh201906** [comment]
I'm now checking why the compatibility test failed on API 26

**Comment by wh201906** [comment]
I found the reason: In API 27 or lower API versions, an activity usually assign the focus to the first focusable view. Sometimes the first focusable view in advanced reminder settings is just the date TextEdit, and you set a `onFocusChangeListener` to it, so the date picker is triggered as the fragment is loaded. The date picker is a model window so the test suite cannot operate other elements then fails. Since API 28, an activity doesn't assign initial focus, so your implementation works.
https://developer.android.com/about/versions/pie/android-9.0-changes-28#focus

I've implemented a fix in f111b18ad0be1bc93c8fcbeffff23981481ee89d but I don't know it it's acceptable for you.

**Comment by wh201906** [comment]
The compatibility test failed for new reason: It cannot find the `editAmount` in the `BasicUITest`
https://github.com/wh201906/medTimer/blob/f111b18ad0be1bc93c8fcbeffff23981481ee89d/app/src/androidTest/java/com/futsch1/medtimer/BasicUITest.java#L83

I cannot reproduce it on my API 26 emulator.

**Comment by Futsch1** [comment]
There's a lot of timing issues in the emulator running on GitHub - it runs just a lot slower than the local one. I found it hard to solve the issues properly other than adding delays, which is no nice way to solve it... The proper solution would be this: https://developer.android.com/training/testing/espresso/idling-resource

I did not have time yet to implement it.

So maybe you try with adding a delay. I will also have a look, but earliest on Monday.

Thank you for your work and help with the compatibility issues!

**Comment by wh201906** [comment]
I managed to record the screen when running the tests.

https://github.com/wh201906/medTimer/blob/f111b18ad0be1bc93c8fcbeffff23981481ee89d/app/src/androidTest/java/com/futsch1/medtimer/BasicUITest.java#L83

***

This recording is for API 34, and the `BasicUITest` starts at 1:03. The flaky code is executed at ~1:20

https://github.com/user-attachments/assets/225ecfa6-b3ce-449f-8e8f-5f135918dd60

<details><summary>Logs:</summary>

> Sun, 22 Dec 2024 18:05:41 GMT > Task :app:dexBuilderDebugAndroidTest
Sun, 22 Dec 2024 18:05:58 GMT Start Recording
Sun, 22 Dec 2024 18:06:07 GMT > Task :app:mergeDebugAndroidTestGlobalSynthetics
Sun, 22 Dec 2024 18:06:07 GMT > Task :app:packageDebug
Sun, 22 Dec 2024 18:06:07 GMT > Task :app:createDebugApkListingFileRedirect
Sun, 22 Dec 2024 18:06:07 GMT > Task :app:processDebugAndroidTestJavaRes NO-SOURCE
Sun, 22 Dec 2024 18:06:08 GMT > Task :app:mergeDebugAndroidTestJniLibFolders
Sun, 22 Dec 2024 18:06:08 GMT > Task :app:checkDebugAndroidTestDuplicateClasses
Sun, 22 Dec 2024 18:06:08 GMT > Task :app:mergeDebugAndroidTestNativeLibs NO-SOURCE
Sun, 22 Dec 2024 18:06:08 GMT > Task :app:mergeDebugAndroidTestJavaResource
Sun, 22 Dec 2024 18:06:08 GMT > Task :app:stripDebugAndroidTestDebugSymbols NO-SOURCE
Sun, 22 Dec 2024 18:06:08 GMT > Task :app:mergeLibDexDebugAndroidTest
Sun, 22 Dec 2024 18:06:08 GMT > Task :app:validateSigningDebugAndroidTest
Sun, 22 Dec 2024 18:06:08 GMT > Task :app:writeDebugAndroidTestSigningConfigVersions
Sun, 22 Dec 2024 18:06:08 GMT > Task :app:mergeProjectDexDebugAndroidTest
Sun, 22 Dec 2024 18:06:09 GMT > Task :app:mergeExtDexDebugAndroidTest
Sun, 22 Dec 2024 18:06:09 GMT > Task :app:packageDebugAndroidTest
Sun, 22 Dec 2024 18:06:09 GMT > Task :app:createDebugAndroidTestApkListingFileRedirect
Sun, 22 Dec 2024 18:06:14 GMT [EmulatorConsole]: Failed to start Emulator console for 5554
Sun, 22 Dec 2024 18:06:27 GMT
Sun, 22 Dec 2024 18:06:27 GMT > Task :app:connectedDebugAndroidTest
Sun, 22 Dec 2024 18:06:27 GMT Starting 8 tests on emulator-5554 - 14
Sun, 22 Dec 2024 18:06:37 GMT
Sun, 22 Dec 2024 18:06:37 GMT emulator-5554 - 14 Tests 0/8 completed. (0 skipped) (0 failed)
Sun, 22 Dec 2024 18:07:07 GMT emulator-5554 - 14 Tests 1/8 completed. (0 skipped) (0 failed)
Sun, 22 Dec 2024 18:07:27 GMT emulator-5554 - 14 Tests 2/8 completed. (0 skipped) (0 failed)
Sun, 22 Dec 2024 18:08:17 GMT emulator-5554 - 14 Tests 3/8 completed. (0 skipped) (0 failed)
Sun, 22 Dec 2024 18:08:37 GMT emulator-5554 - 14 Tests 4/8 completed. (0 skipped) (0 failed)

</details> 

***

This is for API 26, and the `BasicUITest` starts at 0:27. The flaky code is executed at ~0:43

https://github.com/user-attachments/assets/af112609-b0aa-4c4c-ba33-e09612420bd5

<details><summary>Logs:</summary>

> Sun, 22 Dec 2024 18:05:39 GMT > Task :app:mergeExtDexDebugAndroidTest
Sun, 22 Dec 2024 18:05:39 GMT > Task :app:packageDebugAndroidTest
Sun, 22 Dec 2024 18:05:39 GMT > Task :app:createDebugAndroidTestApkListingFileRedirect
Sun, 22 Dec 2024 18:05:41 GMT [EmulatorConsole]: Failed to start Emulator console for 5554
Sun, 22 Dec 2024 18:05:57 GMT
Sun, 22 Dec 2024 18:05:57 GMT > Task :app:connectedDebugAndroidTest
Sun, 22 Dec 2024 18:05:57 GMT Starting 8 tests on emulator-5554 - 8.0.0
Sun, 22 Dec 2024 18:05:59 GMT Start Recording
Sun, 22 Dec 2024 18:06:07 GMT
Sun, 22 Dec 2024 18:06:07 GMT emulator-5554 - 8.0.0 Tests 0/8 completed. (0 skipped) (0 failed)
Sun, 22 Dec 2024 18:06:27 GMT emulator-5554 - 8.0.0 Tests 1/8 completed. (0 skipped) (0 failed)
Sun, 22 Dec 2024 18:06:53 GMT
Sun, 22 Dec 2024 18:06:53 GMT > Task :app:connectedDebugAndroidTest
Sun, 22 Dec 2024 18:06:53 GMT
Sun, 22 Dec 2024 18:06:53 GMT com.futsch1.medtimer.BasicUITest > basicUITest[emulator-5554 - 8.0.0] FAILED 
Sun, 22 Dec 2024 18:06:53 GMT 	java.lang.AssertionError: View did not become visible
Sun, 22 Dec 2024 18:06:53 GMT 	at com.futsch1.medtimer.AndroidTestHelper.onViewWithTimeout(AndroidTestHelper.java:90)
Sun, 22 Dec 2024 18:06:57 GMT
Sun, 22 Dec 2024 18:06:57 GMT emulator-5554 - 8.0.0 Tests 2/8 completed. (0 skipped) (1 failed)
Sun, 22 Dec 2024 18:07:37 GMT emulator-5554 - 8.0.0 Tests 3/8 completed. (0 skipped) (1 failed)
Sun, 22 Dec 2024 18:07:57 GMT emulator-5554 - 8.0.0 Tests 4/8 completed. (0 skipped) (1 failed)

</details> 

***

It looks like the same problem in https://github.com/Futsch1/medTimer/pull/395#issuecomment-2558123124 . I did get the same bug that the date picker appears after I exit from the advanced reminder settings, but I didn't get this bug after the fix (f111b18ad0be1bc93c8fcbeffff23981481ee89d).

**Comment by wh201906** [comment]
My patches in f111b18ad0be1bc93c8fcbeffff23981481ee89d and f078ef3eb2918ce9dcacb41bff95e14800c7cfcb, while functional, are far from elegant. I guess there are better approaches to resolve this but I'm not sure.

**Comment by Futsch1** [comment]
> My patches in f111b18ad0be1bc93c8fcbeffff23981481ee89d and f078ef3eb2918ce9dcacb41bff95e14800c7cfcb, while functional, are far from elegant. I guess there are better approaches to resolve this but I'm not sure.

How do your changes actually impact the behavior of the app on SDK 26? And is the behavior unchanged in the other SDK versions?

By the way, can you also contribute the change to enable recording the emulated tests? This would be helpful to debug future issues.

File: app/src/main/java/com/futsch1/medtimer/medicine/AdvancedReminderSettingsFragment.java [hunk 1]
```diff
@@ -150,10 +150,12 @@ private void setupRemindOnDays() {
     private void setupCycleStartDate() {
         setCycleStartDate(reminder.cycleStartDay);
         editCycleStartDate.setOnFocusChangeListener((v, hasFocus) -> {
-            if (hasFocus) {
-                TimeHelper.DatePickerWrapper datePickerWrapper = new TimeHelper.DatePickerWrapper(requireActivity().getSupportFragmentManager(), R.string.cycle_start_date);
-                datePickerWrapper.show(getCycleStartDate(), this::setCycleStartDate);
-            }
+            v.post(() -> {
```
Comment (Futsch1) [inline app/src/main/java/com/futsch1/medtimer/medicine/AdvancedReminderSettingsFragment.java:156]: Please add a comment here (I know there aren't many in the app, but here I feel it's helpful) to explain why this is posted to the view.

File: app/src/main/res/layout/fragment_advanced_reminder_settings.xml [hunk 2]
```diff
@@ -15,6 +15,8 @@
             android:layout_width="match_parent"
             android:layout_height="wrap_content"
             android:layout_margin="4dp"
+            android:focusable="true"
```
Comment (Futsch1) [inline app/src/main/res/layout/fragment_advanced_reminder_settings.xml:18]: Do you still need this change after the code change posting the listener?

**Review by Futsch1 — CHANGES_REQUESTED** [review]

File: app/src/main/res/layout/fragment_advanced_reminder_settings.xml (hunk 2 above)
Comment (wh201906) [reply app/src/main/res/layout/fragment_advanced_reminder_settings.xml:18]: Yes, both changes are necessary.
- The change in f111b18ad0be1bc93c8fcbeffff23981481ee89d prevents the date picker from being loaded when entering the advanced reminder settings.
- The change in f078ef3eb2918ce9dcacb41bff95e14800c7cfcb ensures the date picker does not load accidentally after exiting the advanced reminder settings.

**Comment by wh201906** [comment]
> How do your changes actually impact the behavior of the app on SDK 26?

For the first fix in commit f111b18ad0be1bc93c8fcbeffff23981481ee89d, I have explained in https://github.com/Futsch1/medTimer/pull/395#issuecomment-2558123124. I'm not sure if I expressed it properly and you can ask me for anything uncertain.
For the second fix in commit f078ef3eb2918ce9dcacb41bff95e14800c7cfcb, I actually didn't figure out why the `onFocusChangeListener` is triggered when exiting the advanced reminder settings. I just postpone the focus check to the end of the event loop, at which point the UI should no longer be in the transition state (probably).

> And is the behavior unchanged in the other SDK versions?

I think the behavior also changes a bit in other SDK versions. 
- The new attributes of the `LinearLayout` makes it focusable, so when a user uses a keyboard to navigate in the advance settings, there will be one more possible element to hold the focus. I think this doesn't affect the UX.
- As for the post delayed focus detection, the delay is minimal (within one event loop iteration) so it's not noticeable to users.

File: app/src/main/java/com/futsch1/medtimer/medicine/AdvancedReminderSettingsFragment.java (hunk 1 above)
Comment (wh201906) [reply app/src/main/java/com/futsch1/medtimer/medicine/AdvancedReminderSettingsFragment.java:156]: Added in 2e9749d8ea8f000afe57c4a739f289489efda855

**Review by Futsch1 — APPROVED** [review]

**Comment by Futsch1** [comment]
Great, PR closed. Thanks for your support, I will look into the recording stuff.

Will release a new version with the lowered SDK requirements soon.

**Comment by Futsch1** [comment]
I am very sorry, but I will have to raise the MinSDK again to 28. I have put so much effort into making the Android tests more stable and succeeded nicely, but I continuously struggled with the devices at SDK 26. When going to 28, it just runs much more stable.

I still have released v1.13.1, which is compatible with SDK 26 and it's just 5 % of total Android devices.

**Comment by wh201906** [comment]
I see. Thanks for your effort.
I guess you could also revert the workaround in this PR because they are not necessary now.

**Pull request merged**